okhttp/src/main/kotlin/okhttp3/internal/tls/DnsUtils.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package okhttp3.internal.tls
+
+/**
+ * A collection of utilities relating to Domain Name System.
+ *
+ * https://raw.githubusercontent.com/apache/httpcomponents-client/master/httpclient5/src/main/java/org/apache/hc/client5/http/utils/DnsUtils.java
+ */
+internal object DnsUtils {
+  /**
+   * This strict interpretation of isUpper for IA5String achieves two things.
+   * Firstly it avoids converting Kelvin (k) to lowercase k.
+   * Secondly it avoids the simple and efficient lowercase ASCII conversion method below
+   * from being applied to other unicode uppercase characters and just moving them to another
+   * undesired unicode character.
+   */
+  private fun isUpper(c: Char): Boolean {
+    return c in 'A'..'Z'
+  }
+
+  // https://tools.ietf.org/html/rfc2459#section-4.2.1.11
+  fun normalizeIA5String(s: String): String {
```

Comment:
In Okio this function is called `ByteString.toAsciiLowercase()` (though to use that one we’d need to roundtrip through a ByteString)